### PR TITLE
[SCRIPT] Pre-trainining script for BERT

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
~requires #499 #500 #501 #503 #504  #489~
~Please review after these PRs are merged, otherwise the diff contains duplicated code changes.~

Summary of changes:
- enhanced training data generation script to produce npz files with multi-processing
- added pre-training script for BERT_base. It currently takes about 8-9 days using fp16 on 8 V100 for training 1M steps from scratch. The produced checkpoint achieves 93% on SST-2, 87.99% on MRPC and 80.99/88.6 on SQuAD 1.1
- added dynamic loss scaling and fp16 trainer, which avoid gradient overflow/underflow using fp16
- modified a few blocks related to BERT to support fp16 dtype, some of them requires explicit casting to avoid loss of precision (e.g. LayerNorm, norm)
- added sample text dataset, and test for training data generation / pre-training on this sample dataset

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/gluonnlp/utils/parameter.py [hunk 1]
```diff
@@ -18,12 +18,89 @@
...
-__all__ = ['clip_grad_global_norm']
+__all__ = ['clip_grad_global_norm', 'grad_global_norm']
 
 import warnings
 
 import numpy as np
 from mxnet import nd
+import mxnet as mx
+
+def grad_global_norm(parameters, max_norm):
```
Comment (szha) [inline src/gluonnlp/utils/parameter.py]: how does it differ from the clip function?

File: scripts/bert/compare_tf_gluon_model.py [hunk 2]
```diff
@@ -32,21 +32,21 @@
 parser = argparse.ArgumentParser(description='Comparison script for BERT model in Tensorflow'
                                              'and that in Gluon')
 parser.add_argument('--input_file', type=str, default='input.txt',
-                    help='sample input file for testing')
+                    help='sample input file for testing, Default is input.txt')
 parser.add_argument('--tf_bert_repo_dir', type=str,
                     default='/home/ubuntu/bert/',
```
Comment (szha) [inline scripts/bert/compare_tf_gluon_model.py]: use `os.path.expanduser('~/...')` instead of assuming a user name

**Comment by mli** [comment]
Job PR-505/28 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-505/28/index.html

File: scripts/bert/compare_tf_gluon_model.py (hunk 2 above)
Comment (szha) [reply scripts/bert/compare_tf_gluon_model.py]: the default is still hard-coded using `/home/ubuntu`

File: src/gluonnlp/model/attention_cell.py [hunk 3]
```diff
@@ -46,7 +47,8 @@ def _masked_softmax(F, att_score, mask):
     """
     if mask is not None:
         # Fill in the masked scores with a very small value
-        att_score = F.where(mask, att_score, -1e18 * F.ones_like(att_score))
+        neg = -10000 if np.dtype(dtype) == np.float16 else -1e18
```
Comment (szhengac) [inline src/gluonnlp/model/attention_cell.py]: Perhaps -1e4?, not sure would integer cause any problem for py2.

File: scripts/bert/fp16_utils.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    """Calculate the 2-norm of gradients of parameters, and how much they should be scaled down
+    such that their 2-norm does not exceed `max_norm`.
+
+    If gradients exist for more than one context for a parameter, user needs to explicitly call
+    ``trainer.allreduce_grads`` so that the gradients are summed first before calculating
+    the 2-norm.
+
+    .. note::
+
+        This function is only for use when `update_on_kvstore` is set to False in trainer.
```
Comment (sravanbabuiitm) [inline scripts/bert/fp16_utils.py]: Good Doc! I have seen a similar question being asked once in questions

File: scripts/bert/fp16_utils.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    # 1 is returned if total_norm does not exceed max_norm.
+    # if scale = NaN or Inf, the result of nd.minimum is undefined. Therefore, we use
+    # choices.take to return NaN or Inf.
+    scale_or_one = nd.maximum(nd.ones((1,), dtype=dtype, ctx=ctx), scale)
+    choices = nd.concat(scale, scale_or_one, dim=0)
+    chosen_scale = choices.take(is_finite)
+    return total_norm, chosen_scale
+
+
+class FP16Trainer(object):
```
Comment (sravanbabuiitm) [inline scripts/bert/fp16_utils.py]: [INFO] why do we instead not use bfloat16 ?
https://en.wikipedia.org/wiki/Bfloat16_floating-point_format
so as to start having parity with Google ?

File: scripts/bert/fp16_utils.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    """Static loss scaler"""
+    def __init__(self, init_scale=1):
+        self.loss_scale = init_scale
+
+    def update_scale(self, overflow):
+        """update loss scale"""
+        pass
+
+class DynamicLossScaler(object):
+    """Class that manages dynamic loss scaling."""
```
Comment (sravanbabuiitm) [inline scripts/bert/fp16_utils.py]: can u add more doc on what dynamic loss scaling means etc. ?

File: scripts/bert/fp16_utils.py (hunk 5 above)
Comment (eric-haibin-lin) [reply scripts/bert/fp16_utils.py]: I don't think CUDA supports bfloat16. At least not now

File: scripts/bert/fp16_utils.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+        self._verbose = verbose
+
+    def update_scale(self, overflow):
+        """dynamically update loss scale"""
+        iter_since_rescale = self._num_steps - self._last_rescale_iter
+        if overflow:
+            self._last_overflow_iter = self._num_steps
+            self._overflows_since_rescale += 1
+            percentage = self._overflows_since_rescale / float(iter_since_rescale)
+            if percentage >= self.tolerance:
```
Comment (sravanbabuiitm) [inline scripts/bert/fp16_utils.py]: Are you scaling only if > 5% of the iterations encounter overflow ?

Comment (eric-haibin-lin) [reply scripts/bert/fp16_utils.py]: Yep. It seems to be the default value used in fairseq.

File: scripts/bert/fp16_utils.py (hunk 6 above)
Comment (eric-haibin-lin) [reply scripts/bert/fp16_utils.py]: Added. Thanks

File: scripts/bert/create_pretraining_data.py [hunk 8]
```diff
@@ -122,84 +155,120 @@ def __str__(self):
...
+
+def print_example(instance, features):
+    logging.debug('*** Example ***')
+    logging.debug('tokens: %s' % ' '.join(instance.tokens))  # pylint: disable=W1201
+
+    for feature_name in features.keys():
+        feature = features[feature_name]
+        logging.debug(  # pylint: disable=W1201
+            '%s: %s' % (feature_name, ' '.join(
+                [str(x) for x in feature])))
```
Comment (szha) [inline scripts/bert/create_pretraining_data.py]: no join is needed when using logging.

File: scripts/bert/create_pretraining_data.py [hunk 9]
```diff
@@ -122,84 +155,120 @@ def __str__(self):
...
-            instance.masked_lm_labels)
-        masked_lm_weights = [1.0] * len(masked_lm_ids)
-
-        while len(masked_lm_positions) < max_predictions_per_seq:
-            masked_lm_positions.append(0)
-            masked_lm_ids.append(0)
-            masked_lm_weights.append(0.0)
+    next_sentence_label = 1 if instance.is_random_next else 0
+
+    features = collections.OrderedDict()
```
Comment (szha) [inline scripts/bert/create_pretraining_data.py]: what's the purpose of using OrderedDict? I can't seem to find an obvious place where order matters.

File: scripts/bert/create_pretraining_data.py [hunk 10]
```diff
@@ -26,36 +26,43 @@
...
+    choices=['numpy', 'recordio'],
+    help='Output file format. If set to "numpy", examples are serialized as a single "npz" file.'
+         'If set to "recordio", examples are serialized using IndexedRecordIO format.')
+
+parser.add_argument(
+    '--vocab',
+    type=str,
+    default=None,
+    help='The dataset name for the vocab file BERT model was trained on. For example, '
+         '"book_corpus_wiki_en_uncased"')
```
Comment (szha) [inline scripts/bert/create_pretraining_data.py]: what are the options? where can they be found

File: scripts/bert/create_pretraining_data.py [hunk 11]
```diff
@@ -436,36 +546,64 @@ def truncate_seq_pair(tokens_a, tokens_b, max_num_tokens, rng):
...
+
+    # prepare workload
+    suffix = 'npz' if args.format == 'numpy' else 'rec'
+    count = 0
+    map_args = []
+    pool_args = (tokenizer, args.max_seq_length, args.dupe_factor,\
+                 args.short_seq_prob, args.masked_lm_prob,
+                 args.max_predictions_per_seq, rng)
+    for i, file_split in enumerate(file_splits):
+        out = os.path.join(output_dir, 'part-%03d.%s'%(i, suffix))
```
Comment (szha) [inline scripts/bert/create_pretraining_data.py]: nit: use str.format going forward as `%` is no longer recommended.

File: scripts/bert/fp16_utils.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+
+    # group norm arrays by ctx
+    def group_by_ctx(arr_list):
+        groups = {}
+        for arr in arr_list:
+            ctx = arr.context
+            if ctx in groups:
+                groups[ctx].append(arr)
+            else:
+                groups[ctx] = [arr]
```
Comment (szha) [inline scripts/bert/fp16_utils.py]: collections.defaultdict can help simplify this to always use append.

File: scripts/bert/fp16_utils.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+        return groups
+    norm_groups = group_by_ctx(norm_arrays)
+
+    # reduce
+    ctx, dtype = arrays[0].context, 'float32'
+    total_norm = mx.nd.zeros((1,), ctx=ctx, dtype=dtype)
+    for group in norm_groups:
+        total_norm += nd.add_n(*norm_groups[group]).as_in_context(ctx)
```
Comment (szha) [inline scripts/bert/fp16_utils.py]: nit: maybe use add_n again to avoid the sequential add?

Comment (szha) [reply scripts/bert/fp16_utils.py]: e.g. `total_norm = nd.add_n(*(nd.add_n(*ng).as_in_context(ctx) for ng in norm_groups.values()))`

File: scripts/bert/fp16_utils.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+        self.fp32_trainer.allreduce_grads()
+        step_size = batch_size * self._scaler.loss_scale
+        if max_norm:
+            norm, ratio = grad_global_norm(self.fp32_trainer._params,
+                                           max_norm * self._scaler.loss_scale)
+            step_size = ratio * step_size
+            self.fp32_trainer.update(step_size)
+            overflow = not np.isfinite(norm.asscalar())
+        else:
+            # TODO(haibin) optimize the performance when max_norm is not present
```
Comment (szha) [inline scripts/bert/fp16_utils.py]: what do you have in mind?

File: scripts/bert/fp16_utils.py [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+            self.fp32_trainer.update(step_size)
+            overflow = self._scaler.has_overflow(self.fp32_trainer._params)
+
+        # update scale based on overflow information
+        self._scaler.update_scale(overflow)
+        if overflow:
+            for param in self.fp32_trainer._params:
+                param.zero_grad()
```
Comment (szha) [inline scripts/bert/fp16_utils.py]: if overflow happens, should the update be skipped?

File: scripts/bert/fp16_utils.py [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+class StaticLossScaler(LossScaler):
+    """Static loss scaler"""
+    def __init__(self, init_scale=1):
+        self.loss_scale = init_scale
+
+    def update_scale(self, overflow):
+        """update loss scale"""
+        pass
+
+class DynamicLossScaler(object):
```
Comment (szha) [inline scripts/bert/fp16_utils.py]: not inheriting LossScaler?

File: scripts/bert/fp16_utils.py [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+            overflow = self._scaler.has_overflow(self.fp32_trainer._params)
+
+        # update scale based on overflow information
+        self._scaler.update_scale(overflow)
+        if overflow:
+            for param in self.fp32_trainer._params:
+                param.zero_grad()
+
+class LossScaler(object):
+    """Abstract loss scaler"""
```
Comment (szha) [inline scripts/bert/fp16_utils.py]: why is update_scale not in the interface?

File: scripts/bert/index.rst [hunk 18]
```diff
@@ -52,5 +52,25 @@ It gets RTE validation accuracy of `70.8% <https://raw.githubusercontent.com/dml
 It gets MNLI validation accuracy ,On dev_matched.tsv: 84.6%
 On dev_mismatched.tsv: 84.7%. `log <https://github.com/dmlc/web-data/blob/master/gluonnlp/logs/bert/finetuned_mnli.log>`_
 
-
 Some other tasks can be modeled with `--task_name` parameter.
+
+Pre-training with BERT
+~~~~~~~~~~~~~~~~~~~~~~
+
```
Comment (szha) [inline scripts/bert/index.rst]: remember to report the downstream task results from finetuning the pre-trained model from this.

File: scripts/tests/test_bert.py [hunk 19]
```diff
@@ -0,0 +1,39 @@
+import os
+import subprocess
+import time
+import mxnet as mx
+
+import pytest
+
+@pytest.mark.serial
+@pytest.mark.remote_required
+def test_pretrain():
```
Comment (szha) [inline scripts/tests/test_bert.py]: ideally this should go into test_scripts. does it only work with mxnet nightly?

File: src/gluonnlp/metric/masked_accuracy.py [hunk 20]
```diff
@@ -249,7 +249,10 @@ def update(self, labels, preds, masks=None):
 
         for label, pred_label, mask in zip(labels, preds, masks):
             if pred_label.shape != label.shape:
-                pred_label = ndarray.argmax(pred_label, axis=self.axis)
+                # TODO(haibin) topk does not support fp16. Issue tracked at:
+                # https://github.com/apache/incubator-mxnet/issues/14125
```
Comment (szha) [inline src/gluonnlp/metric/masked_accuracy.py]: also leave a note that you're using topk due to speed issue in argmax, and link to the issue `https://github.com/apache/incubator-mxnet/issues/11061`

File: scripts/tests/test_bert.py (hunk 19 above)
Comment (eric-haibin-lin) [reply scripts/tests/test_bert.py]: Yes

File: scripts/bert/fp16_utils.py [hunk 21]
```diff
@@ -18,6 +18,7 @@
 
 """Trainer for mixed precision training."""
 import logging
+from collections import defaultdict
```
Comment (szha) [inline scripts/bert/fp16_utils.py]: import collections. then use collections.defaultdict

File: scripts/bert/fp16_utils.py [hunk 22]
```diff
@@ -79,22 +80,17 @@ def _norm(array):
...
         return groups
     norm_groups = group_by_ctx(norm_arrays)
 
     # reduce
     ctx, dtype = arrays[0].context, 'float32'
-    total_norm = mx.nd.zeros((1,), ctx=ctx, dtype=dtype)
-    for group in norm_groups:
-        total_norm += nd.add_n(*norm_groups[group]).as_in_context(ctx)
-    total_norm = nd.sqrt(total_norm)
+    norms = [nd.add_n(*norm_groups[g]).as_in_context(ctx) for g in norm_groups]
```
Comment (szha) [inline scripts/bert/fp16_utils.py]: why not iterate on values directly? (i.e. `[...for g in norm_groups.values()]`)

File: scripts/bert/create_pretraining_data.py (hunk 10 above)
Comment (eric-haibin-lin) [reply scripts/bert/create_pretraining_data.py]: options added

**Comment by mli** [comment]
Job PR-505/46 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-505/46/index.html

File: scripts/bert/fp16_utils.py [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+class LossScaler(object):
+    """Abstract loss scaler"""
+    def has_overflow(self, params):
+        """ detect inf and nan """
+        is_not_finite = 0
+        for param in params:
+            if param.grad_req != 'null':
+                grad = param.list_grad()[0]
+                is_not_finite += mx.nd.contrib.isnan(grad).sum()
+                is_not_finite += mx.nd.contrib.isinf(grad).sum()
```
Comment (szha) [inline scripts/bert/fp16_utils.py]: early return?

Comment (szha) [reply scripts/bert/fp16_utils.py]: never mind, it would likely cause too many synchronizations.

**Comment by mli** [comment]
Job PR-505/47 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-505/47/index.html
